k2/csrc/ragged_utils.h

```diff
@@ -0,0 +1,172 @@
+
+  This function may do a quick or more thorough check depending on the debug
+  level k2 was compiled with it is intended for situations where it's required
+  for the axes to be the same and the program needs to exit or raise an
+  exception if not.
+
+  Note: you can use the function Equal(RaggedShape&, RaggedShape&) if you
+  want to check the return status directly.
+
+    @param [in] num_srcs   Length of the list of sources.
+    @param [in] axis       Axis to check, e.g. 0; must satisfy
+                           `0 <= axis < src[0]->NumAxes() - 1.`
+    @param [in] srcs      The sources to check
+
+ */
+void CheckAxisEqual(int32_t num_srcs,
+                    int32_t axis,
+                    RaggedShape **src);
+
+
+/*
+   Stacking operation on a single axis of an array of RaggedShape sources,
+   that uses the order "all rows for source 0" then "all rows for source 1",
+   etc.,
+
+   (Check the usage message carefully, because this interface is not very
+   intuitive).
+
+      @param [in] num_srcs   Number of sources to append/merge (size of array `src`)
+      @param [in] axis       Axis to operate on, viewed as index into src[i]->Axes().
+                             Must satisfy 0 <= axis < src[0]->Axes()  - 1.
+      @param [in] src        Array of sources, must have the same device and num-axes.
+      @param [out,optional]  merge_map    If not nullptr, will be set to an array
+                             that indicates the source of each element on axis
+                             `axis+1`, with merge_map->Dim() == the sum of
+                             src[i]->TotSize(axis+1).  If `m = (*merge_map)[i]`,
+                             then `m % num_srcs` indicates the source for this
+                             item and `m / num_srcs` indicates the position of this
+                             item within its source.
+
+      @return               Return a RaggedShape with `NumAxes() == 2`, i.e. `Axes().size() == 1`,
+                            that is the result of appending the sources together; its
+                            TotSize(0) will be the sum of src[i]->TotSize(axis),
+                            and its TotSize(1) will be the sum of src[i]->TotSize(axis+1).
+                            Its order will be: all elements of *src[0], all elements of
+                            *src[1], and so on.
+
+    EXAMPLE: suppose num_srcs == 2, and axis == 0, and src[0]->NumAxes() == 2.
+    And suppose *src[0] == [ x x x ] [ x ] and *src[1] = [ ] [ x x ].  Then
+    ans == [ x x x ] [ x ] [ ] [ x x ], and merge_map (if requested) will
+    be [ 0 2 4 6 1 3 ].
+ */
+RaggedShape AppendRaggedAxisBefore(int32_t num_srcs,
+                                   int32_t axis,
+                                   RaggedShape **src,
+                                   Array1<int32_t> *merge_map = nullptr);
+
+
+
+/*
+   Intersperses rows from a single layer of an array of RaggedShape sources,
```

Comment:
@danpovey, just a question, this function will be called in `MergeRaggedAxis`? Anyway, it's glad to see your these declarations, I was thinking `Append` on other axis would be very hard.

Reply:
It is a primitive to be used in Append() and Stack() operations.